File: sort/merge_sort.java
public class merge_sort {
    public static void main(String[] args) {
        /*
        Merge sort - Merge sort uses recursion to divide an array into subarrays 
        and sorts them based on the smallest subarrays possible.
        For more info, see README.

        Runtime Complexity - O(n log n)
        Space - O(n)
         */

        // Input specification, this can be any array
        int[] array = {8, 2, 5, 3, 4, 7, 6, 1};

        mergeSort(array);

        for (int i : array) {
            System.out.print(i + " ");
        }
        
    }

    private static void mergeSort(int[] array) {
        // Recursive Merge Sort function
        
        int length = array.length;
		if (length <= 1) return; //base case
		
		int middle = length / 2;
		int[] leftArray = new int[middle];
		int[] rightArray = new int[length - middle];
		
		int i = 0; //left array
		int j = 0; //right array
		
		for (; i < length; i++) {
			if(i < middle) {
				leftArray[i] = array[i];
			}
			else {
				rightArray[j] = array[i];
				j++;
			}
		}
		mergeSort(leftArray);
		mergeSort(rightArray);
		merge(leftArray, rightArray, array);
    }

    private static void merge(int[] leftArray, int[] rightArray, int[] array) {
        int leftSize = array.length / 2;
		int rightSize = array.length - leftSize;
		int i = 0, l = 0, r = 0; //indices
		
		//check the conditions for merging
		while(l < leftSize && r < rightSize) {
			if(leftArray[l] < rightArray[r]) {
				array[i] = leftArray[l];
				i++;
				l++;
			}
			else {
				array[i] = rightArray[r];
				i++;
				r++;
			}
		}
		while(l < leftSize) {
			array[i] = leftArray[l];
			i++;
			l++;
		}
		while(r < rightSize) {
			array[i] = rightArray[r];
			i++;
			r++;
		}
    }
}

File: sort/merge_sort.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> merge(vector<int>& list1, vector<int>& list2) {
    vector<int> merged_list;

    int i = 0, j = 0;

    while (i < list1.size() && j < list2.size()) {
        if (list1[i] < list2[j]) {
            merged_list.push_back(list1[i]);
            i++;
        } else {
            merged_list.push_back(list2[j]);
            j++;
        }
    }

    if (i < list1.size()) {
        for (int z = i; z < list1.size(); z++) {
            merged_list.push_back(list1[z]);
        }
    } else {
        for (int z = j; z < list2.size(); z++) {
            merged_list.push_back(list2[z]);
        }
    }

    return merged_list;
}

vector<int> mergeSort(vector<int>& list) {
    if (list.size() <= 1) {
        return list;
    }

    int mid = list.size() / 2;

    vector<int> left(list.begin() + 0, list.begin() + mid);
    vector<int> right(list.begin() + mid, list.end());

    left = mergeSort(left);
    right = mergeSort(right);

    return merge(left, right);
}

int main() {
    vector<int> list = {64, 34, 25, 12, 22, 11, 90, 5};

    list = mergeSort(list);

    for (int i = 0; i < list.size(); i++) cout << list[i] << "\n";

    return 0;
}
